libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1239 @@
+                         eos_discovery_feed_knowledge_app_proxy_get_knowledge_search_object_path (data->ka_proxy),
+                         eos_discovery_feed_knowledge_app_proxy_get_knowledge_app_id (data->ka_proxy),
+                         data->direction ? data->direction : EOS_DISCOVERY_FEED_CARD_LAYOUT_DIRECTION_IMAGE_FIRST,
+                         data->thumbnail_size);
+      g_ptr_array_add (orderable_stores,
+                       eos_discovery_feed_orderable_model_new (EOS_DISCOVERY_FEED_BASE_CARD_STORE (store),
+                                                               data->type,
+                                                               eos_discovery_feed_knowledge_app_proxy_get_desktop_id (data->ka_proxy)));
+    }
+
+    return g_steal_pointer (&orderable_stores);
+}
+
+typedef struct _AppendDiscoveryFeedContentFromProxyData
+{
+  gchar                                            *method;
+  EosDiscoveryFeedCardStoreType                     type;
+  EosDiscoveryFeedCardLayoutDirection               direction;
+  EosDiscoveryFeedKnowledgeAppCardStoreFactoryFunc  factory;
+  guint                                             thumbnail_size;
+} AppendDiscoveryFeedContentFromProxyData;
+
+static AppendDiscoveryFeedContentFromProxyData *
+append_discovery_feed_content_from_proxy_data_new (const gchar                                     *method,
+                                                   EosDiscoveryFeedCardStoreType                    type,
+                                                   EosDiscoveryFeedCardLayoutDirection              direction,
+                                                   EosDiscoveryFeedKnowledgeAppCardStoreFactoryFunc factory,
+                                                   guint                                            thumbnail_size)
+
+{
+  AppendDiscoveryFeedContentFromProxyData *data = g_new0 (AppendDiscoveryFeedContentFromProxyData, 1);
+
+  data->method = g_strdup (method);
+  data->type = type;
+  data->direction = direction;
+  data->factory = factory;
+  data->thumbnail_size = thumbnail_size;
+
+  return data;
+}
+
+static void
+append_discovery_feed_content_from_proxy_data_free (AppendDiscoveryFeedContentFromProxyData *data)
+{
+  g_clear_pointer (&data->method, g_free);
+
+  g_free (data);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (AppendDiscoveryFeedContentFromProxyData,
+                               append_discovery_feed_content_from_proxy_data_free);
+
+static gpointer
+append_discovery_feed_content_from_proxy (EosDiscoveryFeedKnowledgeAppProxy  *ka_proxy,
+                                          gpointer                            proxy_data,
+                                          GCancellable                       *cancellable,
+                                          GError                            **error)
+{
+  g_autoptr(AppendDiscoveryFeedContentFromProxyData) data = proxy_data;
```

Comment:
Convention I've been using has been to take ownership of the closure inside the callback, let me know if it is better to have the caller be passed a GDestroyNotify and take ownership from there instead.